File: flutter/lib/src/micropay_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;

import 'models/payment_intent.dart';
import 'models/micropay_error.dart';

/// Configuration options for the Micropay client.
class MicropayConfig {
  /// Your Micropay public key (pk_test_... or pk_live_...).
  final String publicKey;

  /// API base URL. Defaults to production.
  final String baseUrl;

  /// Request timeout duration.
  final Duration timeout;

  /// Enable debug logging.
  final bool debug;

  MicropayConfig({
    required this.publicKey,
    this.baseUrl = 'https://api.micropay.io/v1',
    this.timeout = const Duration(seconds: 30),
    this.debug = false,
  });

  /// Whether this is a test environment.
  bool get isTestMode => publicKey.startsWith('pk_test_');
}

/// The main Micropay client for Flutter applications.
///
/// Use this class to create and manage payment intents.
///
/// Example:
/// ```dart
/// final micropay = Micropay(publicKey: 'pk_test_...');
///
/// final intent = await micropay.createPaymentIntent(
///   amount: 500,
///   phoneNumber: '254712345678',
/// );
/// ```
class Micropay {
  final MicropayConfig _config;
  final http.Client _httpClient;

  /// Creates a new Micropay client.
  ///
  /// [publicKey] is your Micropay public key.
  /// [baseUrl] optionally overrides the API base URL.
  /// [timeout] sets the request timeout (default: 30 seconds).
  /// [debug] enables debug logging.
  Micropay({
    required String publicKey,
    String? baseUrl,
    Duration? timeout,
    bool debug = false,
    http.Client? httpClient,
  })  : _config = MicropayConfig(
          publicKey: publicKey,
          baseUrl: baseUrl ?? 'https://api.micropay.io/v1',
          timeout: timeout ?? const Duration(seconds: 30),
          debug: debug,
        ),
        _httpClient = httpClient ?? http.Client();

  /// The current configuration.
  MicropayConfig get config => _config;

  /// Whether the client is in test mode.
  bool get isTestMode => _config.isTestMode;

  /// Creates a new payment intent.
  ///
  /// [amount] is the amount in the smallest currency unit (e.g., cents).
  /// [phoneNumber] is the M-Pesa phone number to charge.
  /// [currency] is the three-letter currency code (default: 'KES').
  /// [description] is an optional description for the payment.
  /// [metadata] is optional key-value data to attach to the payment.
  /// [idempotencyKey] prevents duplicate charges for the same request.
  Future<PaymentIntent> createPaymentIntent({
    required int amount,
    required String phoneNumber,
    String currency = 'KES',
    String? description,
    Map<String, dynamic>? metadata,
    String? idempotencyKey,
  }) async {
    final body = {
      'amount': amount,
      'currency': currency,
      'phone_number': phoneNumber,
      if (description != null) 'description': description,
      if (metadata != null) 'metadata': metadata,
    };

    final response = await _post(
      '/payment-intents',
      body: body,
      idempotencyKey: idempotencyKey,
    );

    return PaymentIntent.fromJson(response);
  }

  /// Confirms a payment intent.
  ///
  /// This initiates the M-Pesa STK Push to the user's phone.
  ///
  /// [paymentIntentId] is the ID of the payment intent to confirm.
  /// [clientSecret] is the client secret from the payment intent.
  Future<PaymentIntent> confirmPaymentIntent(
    String paymentIntentId,
    String clientSecret,
  ) async {
    final response = await _post(
      '/payment-intents/$paymentIntentId/confirm',
      body: {'client_secret': clientSecret},
    );

    return PaymentIntent.fromJson(response);
  }

  /// Retrieves a payment intent by ID.
  ///
  /// [paymentIntentId] is the ID of the payment intent to retrieve.
  Future<PaymentIntent> getPaymentIntent(String paymentIntentId) async {
    final response = await _get('/payment-intents/$paymentIntentId');
    return PaymentIntent.fromJson(response);
  }

  /// Polls for payment completion.
  ///
  /// Returns the final payment intent when the payment succeeds or fails.
  ///
  /// [paymentIntentId] is the ID of the payment intent to poll.
  /// [pollInterval] is the time between polls (default: 2 seconds).
  /// [maxAttempts] is the maximum number of poll attempts (default: 60).
  /// [onStatusChange] is called when the status changes.
  Future<PaymentIntent> pollPaymentStatus(
    String paymentIntentId, {
    Duration pollInterval = const Duration(seconds: 2),
    int maxAttempts = 60,
    void Function(PaymentIntentStatus status)? onStatusChange,
  }) async {
    PaymentIntentStatus? lastStatus;

    for (var attempt = 0; attempt < maxAttempts; attempt++) {
      final intent = await getPaymentIntent(paymentIntentId);

      if (intent.status != lastStatus) {
        lastStatus = intent.status;
        onStatusChange?.call(intent.status);
      }

      if (intent.status.isTerminal) {
        return intent;
      }

      await Future.delayed(pollInterval);
    }

    throw MicropayError.timeout();
  }

  /// Cancels a payment intent.
  ///
  /// [paymentIntentId] is the ID of the payment intent to cancel.
  Future<PaymentIntent> cancelPaymentIntent(String paymentIntentId) async {
    final response = await _post('/payment-intents/$paymentIntentId/cancel');
    return PaymentIntent.fromJson(response);
  }

  // HTTP helper methods

  Future<Map<String, dynamic>> _get(String path) async {
    final uri = Uri.parse('${_config.baseUrl}$path');
    
    _log('GET $uri');

    try {
      final response = await _httpClient
          .get(uri, headers: _headers)
          .timeout(_config.timeout);

      return _handleResponse(response);
    } on TimeoutException {
      throw MicropayError.timeout();
    } catch (e) {
      throw MicropayError.network('Failed to connect to Micropay', cause: e);
    }
  }

  Future<Map<String, dynamic>> _post(
    String path, {
    Map<String, dynamic>? body,
    String? idempotencyKey,
  }) async {
    final uri = Uri.parse('${_config.baseUrl}$path');
    
    _log('POST $uri');
    if (body != null) _log('Body: $body');

    final headers = Map<String, String>.from(_headers);
    if (idempotencyKey != null) {
      headers['Idempotency-Key'] = idempotencyKey;
    }

    try {
      final response = await _httpClient
          .post(
            uri,
            headers: headers,
            body: body != null ? jsonEncode(body) : null,
          )
          .timeout(_config.timeout);

      return _handleResponse(response);
    } on TimeoutException {
      throw MicropayError.timeout();
    } catch (e) {
      throw MicropayError.network('Failed to connect to Micropay', cause: e);
    }
  }

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Authorization': 'Bearer ${_config.publicKey}',
        'X-Micropay-Client': 'flutter-sdk/0.1.0',
      };

  Map<String, dynamic> _handleResponse(http.Response response) {
    _log('Response ${response.statusCode}: ${response.body}');

    Map<String, dynamic> data;
    try {
      data = jsonDecode(response.body) as Map<String, dynamic>;
    } catch (e) {
      throw MicropayError(
        code: 'invalid_response',
        message: 'Failed to parse response from Micropay',
        statusCode: response.statusCode,
      );
    }

    if (response.statusCode >= 200 && response.statusCode < 300) {
      return data;
    }

    throw MicropayError.fromJson(data, statusCode: response.statusCode);
  }

  void _log(String message) {
    if (_config.debug) {
      // ignore: avoid_print
      print('[Micropay] $message');
    }
  }

  /// Disposes the HTTP client.
  void dispose() {
    _httpClient.close();
  }
}

File: react-native/src/MicropayClient.js
/**
 * MicropayClient - API Client for React Native
 * 
 * Handles all HTTP communication with the Micropay API.
 */

const DEFAULT_API_URL = 'https://micropay.dev';

export class MicropayClient {
    constructor(config) {
        this.publicKey = config.publicKey;
        this.secretKey = config.secretKey;
        this.apiUrl = config.apiUrl || DEFAULT_API_URL;
        this.environment = config.environment || 'sandbox';
    }

    /**
     * Get the appropriate API key for the request type
     */
    _getApiKey(requireSecret = false) {
        if (requireSecret) {
            if (!this.secretKey) {
                throw new Error('Secret key required for this operation. Use sk_test_... or sk_live_...');
            }
            return this.secretKey;
        }
        return this.publicKey || this.secretKey;
    }

    /**
     * Make an API request
     */
    async _request(endpoint, options = {}) {
        const url = endpoint.startsWith('http') ? endpoint : `${this.apiUrl}/${endpoint}`;
        const apiKey = this._getApiKey(options.requireSecret);

        const headers = {
            'Content-Type': 'application/json',
            'x-api-key': apiKey,
            ...options.headers
        };

        if (options.idempotencyKey) {
            headers['Idempotency-Key'] = options.idempotencyKey;
        }

        const response = await fetch(url, {
            method: options.method || 'GET',
            headers,
            body: options.body ? JSON.stringify(options.body) : undefined
        });

        const data = await response.json();

        if (!response.ok) {
            throw new MicropayError(data.error || 'Request failed', response.status, data);
        }

        return data;
    }

    /**
     * Create a Payment Intent (Server-side operation)
     */
    async createPaymentIntent({ amount, currency = 'KES', customerPhone, description, metadata }) {
        return this._request('v1/payment_intents', {
            method: 'POST',
            requireSecret: true,
            body: {
                amount,
                currency,
                customer_phone: customerPhone,
                description,
                metadata
            }
        });
    }

    /**
     * Confirm a Payment Intent (Client-side operation)
     * Triggers the M-Pesa STK Push to the customer's phone
     */
    async confirmPaymentIntent(intentId, { clientSecret, phoneNumber }) {
        return this._request(`v1/payment_intents/${intentId}/confirm`, {
            method: 'POST',
            requireSecret: false,
            body: {
                client_secret: clientSecret,
                phone_number: phoneNumber
            }
        });
    }

    /**
     * Get Payment Intent status
     */
    async getPaymentIntent(intentId) {
        return this._request(`v1/payment_intents/${intentId}`, {
            method: 'GET',
            requireSecret: false
        });
    }

    /**
     * Poll for payment completion
     * @param {string} intentId - Payment Intent ID
     * @param {object} options - Polling options
     * @returns {Promise<object>} Final payment status
     */
    async pollPaymentStatus(intentId, options = {}) {
        const maxAttempts = options.maxAttempts || 12;
        const intervalMs = options.intervalMs || 5000;
        const onStatusChange = options.onStatusChange || (() => { });

        for (let attempt = 0; attempt < maxAttempts; attempt++) {
            const intent = await this.getPaymentIntent(intentId);

            onStatusChange(intent.status, attempt);

            if (intent.status === 'succeeded' || intent.status === 'failed') {
                return intent;
            }

            // Wait before next poll
            await new Promise(resolve => setTimeout(resolve, intervalMs));
        }

        // Timeout - return last known state
        return this.getPaymentIntent(intentId);
    }
}

/**
 * Custom Error class for Micropay errors
 */
export class MicropayError extends Error {
    constructor(message, statusCode, data) {
        super(message);
        this.name = 'MicropayError';
        this.statusCode = statusCode;
        this.data = data;
    }
}
